src/analyzer/protocol/websocket/websocket.spicy

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) 2024 by the Zeek Project. See LICENSE for details.
+
+module WebSocket;
+
+import spicy;
+import zeek;
+
+const OPCODE_CONTINUATION = 0x00;
+const OPCODE_TEXT         = 0x01;
+const OPCODE_BINARY       = 0x02;
+const OPCODE_CLOSE        = 0x08;
+const OPCODE_PING         = 0x09;
+const OPCODE_PONG         = 0x0a;
+
+# Are bytes copied like a string in C++, or is that ref counted and cheap
+# like it would be in Python?
+function unmask(inout f: Frame, chunk: bytes): bytes {
+  if ( ! f.mask )
+    return chunk;
+
+  local unmasked_chunk: bytes;
+  local i = 0;
+  # This is very very expensive :-( Creates a new shared_ptr for
+  # each byte we iterate over it seems. Visible in the first
+  # flamegraph I looked at.
+  for ( c in chunk ) {
```

Comment:
This iteration is done using safe iterators on the underlying bytes object, instantiating shared_ptrs over and over. It's by far the dominating perf sink here. 70% of samples fall into shared_ptr constructor and destructors and it's 30% of the total runtime for my testing.

![Screenshot from 2024-01-24 20-45-24](https://github.com/zeek/zeek/assets/2100509/1f192e9c-8607-41ce-9c26-0d3098641516)

Reply:
We discussed this at length already. The issue here is that iteration over a `bytes` container is by design safe in Spicy, i.e., even in a for each loop we still check for element validity. This is needed since the loop body could invalidate elements (also the reason a `for_each` method taking a callback on `bytes` could not avoid that).

Like discussed, if performance of this piece is critical one approach would be to move it to a C++ function; if this code wouldn't be called during `&chunked` parsing it very likely would also be possible to more directly express this in the parsing code.

Reply:
@bbannier - I've implemented unmask in C++ and just doing that improved overall runtime by 1.8x (from 22 down to 12 seconds). Mind checking out the unmask.cc file and signing-off if it doesn't looks too troublesome?